lib/pos-parser/nfa.ts

```diff
@@ -43,27 +47,113 @@ class State {
     }
 }
 
+interface Token {
+    token : string,
+    pos : string|undefined
+}
+
+function arrayMatch(a : string[], b : string[]) : number {
+    return a.findIndex((element, i) => {
+        for (let j = 0; j < b.length; j++) {
+            if (i + j > a.length - 1)
+                return false;
+            if (a[i + j] !== b[j])
+                return false;
+        }
+        return true;
+    });
+}
+
+function arrayMatchCount(a : string[], b : string[]) : number {
+    let count = 0;
+    for (let i = 0; i < a.length - b.length + 1; i++) {
+        let match = true;
+        for (let j = 0; j < b.length; j++) {
+            if (a[i + j] !== b[j])
+                match = false;
+        }
+        if (match)
+            count += 1;
+    }
+    return count;
+}
+
 export class NFA {
```

Comment:
At this point, NFA is not just a generic NFA engine, it's specifically for your canonical form parser. Maybe split this into two classes, one that just implements the NFA and another that has the preprocessing and POS logic?